twitterscraper/tweet.py

```diff
@@ -6,61 +6,135 @@
-        self.is_retweet = is_retweet
-        self.retweeter_username = retweeter_username
-        self.retweeter_userid = retweeter_userid
-        self.retweet_id = retweet_id
+        # tweet text
         self.text = text
-        self.html = html
+        self.text_html = text_html
+        self.links = links
+        self.hashtags = hashtags
+        # tweet media
+        self.has_media = has_media
+        self.img_urls = img_urls
+        self.video_url = video_url
+        # tweet actions numbers
+        self.likes = likes
+        self.retweets = retweets
+        self.replies = replies
+        self.is_replied = is_replied
+        # detail of reply to others
+        self.is_reply_to = is_reply_to
+        self.parent_tweet_id = parent_tweet_id
+        self.reply_to_users = reply_to_users
 
     @classmethod
     def from_soup(cls, tweet):
         tweet_div = tweet.find('div', 'tweet')
-        username = tweet_div["data-screen-name"]
-        fullname = tweet_div["data-name"]
+
+        # user name & id
+        screen_name = tweet_div["data-screen-name"].strip('@')
+        username = tweet_div["data-name"]
         user_id = tweet_div["data-user-id"]
-        tweet_id = tweet_div["data-tweet-id"]
+
+        # tweet basic data
+        tweet_id = tweet_div["data-tweet-id"]  # equal to 'data-item-id'
         tweet_url = tweet_div["data-permalink-path"]
         timestamp_epochs = int(tweet.find('span', '_timestamp')['data-time'])
         timestamp = datetime.utcfromtimestamp(timestamp_epochs)
-        try:
-            retweet_id = tweet_div["data-retweet-id"]
-            retweeter_username = tweet_div["data-retweeter"]
-            retweeter_userid = tweet_div.find('a', "pretty-link js-user-profile-link")["data-user-id"]
-            is_retweet = 1
-        except:
-            retweet_id = ""
-            retweeter_username = ""
-            retweeter_userid = ""
-            is_retweet = 0
-
-        text = tweet.find('p', 'tweet-text').text or ""
-        replies = int(tweet.find(
-            'span', 'ProfileTweet-action--reply u-hiddenVisually').find(
+
+        # tweet text
+        soup_html = tweet_div \
+            .find('div', 'js-tweet-text-container') \
+            .find('p', 'tweet-text')
+        text_html = str(soup_html) or ""
+        text = soup_html.text or ""
+        links, hashtags = cls.extract_links_and_hashtags(soup_html)
+
+        # tweet media
+        # --- imgs
+        soup_imgs = tweet_div.find_all('div', 'AdaptiveMedia-photoContainer')
+        img_urls = [
+            img['data-image-url'] for img in soup_imgs
+        ] if soup_imgs else []
+
+        # --- videos
+        video_div = tweet_div.find('div', 'PlayableMedia-container')
+        video_url = video_div.find('a')['href'] if video_div else ''
+        has_media = True if img_urls or video_url else False
+        # eliminate video_url from links for duplicate
+        links = [link for link in links if link != video_url]
```

Comment:
I think it can be quite confusing to use link five times in one line. An alternative could be to use `filter`:
`links = list(filter(lambda x: x != video_url, links)`)

Reply:
Oh, this is simple way!
I'll rewrite.